File: shared/protocol/catalog.ts
import { z } from "zod";

/**
 * The module catalog the engine prints with `--catalog`: every registered module's ports, params, ranges,
 * units and enum labels, plus the conventions a patch is written in.
 *
 * These schemas are the renderer's only description of what a module is. Nothing here is hand-maintained
 * per module: adding a module to the engine changes the JSON, not this file. What this file does maintain
 * is the *shape* of that JSON, so a mismatch between the two surfaces as a parse error at startup rather
 * than as an undefined somewhere deep in the patch editor.
 */

/** Continuous ports carry a signal every frame; event ports carry notes and other discrete messages. */
export const PortKindSchema = z.enum(["continuous", "event"]);

/**
 * What a port means, for colouring and for the editor's connection hints only -- the engine never rejects a
 * patch over a role mismatch. Extending this list is how a new kind of signal is introduced.
 *
 * `note` is a role, not a kind: it marks an event port carrying a stream of pitch and velocity, where an
 * event port carrying bare triggers stays `gate`. Each role has a `--color-signal-<role>` token.
 */
export const SignalRoleSchema = z.enum([
  "any",
  "audio",
  "cv",
  "gate",
  "pitch",
  "phase",
  "note",
]);

export const ParamUnitSchema = z.enum([
  "none",
  "hz",
  "seconds",
  "db",
  "semitones",
  "percent",
  "ratio",
]);

export const ParamCurveSchema = z.enum(["linear", "log", "exp"]);

export const UiWidgetSchema = z.enum(["slider", "knob", "toggle", "select"]);

/**
 * One port. `implicit` marks the port the engine adds for a modulatable param (`param:<id>`): the editor
 * draws it on the knob rather than in the port list, and `param` says which knob.
 */
export const PortDescSchema = z
  .object({
    id: z.string().min(1),
    name: z.string(),
    kind: PortKindSchema,
    role: SignalRoleSchema,
    doc: z.string(),
    implicit: z.boolean(),
    param: z.string().optional(),
  })
  .strict()
  .refine((p) => p.implicit === (p.param !== undefined), {
    message:
      "an implicit port names the param it feeds, and only an implicit port does",
  });

export const ParamFlagsSchema = z
  .object({
    /** Takes a modulation signal; the engine gives it an implicit `param:<id>` input port. */
    modulatable: z.boolean(),
    integer: z.boolean(),
    enum: z.boolean(),
    hidden: z.boolean(),
    noSmooth: z.boolean(),
    /** Cannot be changed on a live node: the engine rebuilds the node instead. Never modulatable. */
    structural: z.boolean(),
    /**
     * The module wants this control on its face.
     *
     * Which handful of a module's parameters matter is the module's knowledge: nothing about a
     * parameter table says a filter's cutoff is reached for more often than its formant spread. So the
     * engine says, and the interface shows these without being asked.
     */
    primary: z.boolean(),
  })
  .strict();

export const ParamDescSchema = z
  .object({
    id: z.string().min(1),
    name: z.string(),
    min: z.number(),
    max: z.number(),
    default: z.number(),
    unit: ParamUnitSchema,
    curve: ParamCurveSchema,
    flags: ParamFlagsSchema,
    enumLabels: z.array(z.string()).min(1).optional(),
    uiWidget: UiWidgetSchema,
    group: z.string().optional(),
    doc: z.string(),
  })
  .strict()
  .refine((p) => p.min < p.max, { message: "param needs min < max" })
  .refine((p) => p.default >= p.min && p.default <= p.max, {
    message: "param default is outside its own range",
  })
  .refine((p) => !p.flags.enum || p.enumLabels !== undefined, {
    message: "an enum param carries its labels",
  })
  .refine((p) => !(p.flags.structural && p.flags.modulatable), {
    message: "a structural param can never be modulatable",
  });

/**
 * A string a module owns: a note pattern, a sample path, an expression.
 *
 * Params are numbers everywhere, because a number can be smoothed, modulated and swept. A string
 * can do none of those, so it lives in the node's `data` under this id and is *structural* -- the
 * engine rebuilds the node when it changes. Declaring it here is what lets the interface generate
 * an editor for it, the same way it generates a knob from a param.
 */
export const TextDescSchema = z
  .object({
    id: z.string().min(1),
    name: z.string(),
    default: z.string(),
    flags: z.object({ multiline: z.boolean() }).strict(),
    /** Which editor mode to open. Absent, or a name we do not know, is plain text. */
    language: z.string().optional(),
    placeholder: z.string().optional(),
    doc: z.string(),
  })
  .strict();

export const ModuleFlagsSchema = z
  .object({
    /** Writes to the engine output rather than to a port of its own. */
    terminal: z.boolean(),
    needsTransport: z.boolean(),
    writesTelemetry: z.boolean(),
    /**
     * Can draw one cycle of itself: `module.preview` returns it. The editor gives such a module a wave
     * panel on its face and asks the engine what to put in it, so the picture is computed from the
     * same parameters the sound is, by the module that makes the sound.
     */
    previewsWave: z.boolean(),
    /**
     * Its telemetry slot carries a rolling window of the signal it is fed (`TelemetryKind.Scope`).
     * The editor gives such a module a scope panel on its face and draws the window there, read from
     * the engine's segment at frame rate.
     */
    publishesScope: z.boolean(),
    /**
     * Its telemetry slot carries the last value on its input, per channel (`TelemetryKind.Value`).
     * The editor gives such a module a readout on its face and writes the number there.
     */
    publishesValue: z.boolean(),
    /**
     * Its telemetry slot carries the level on its input: held peak, RMS and a clip flag per channel
     * (`TelemetryKind.Meter`). The editor gives such a module a level meter on its face.
     */
    publishesMeter: z.boolean(),
    /** Publishes the notes it is playing, which is what a piano roll on its face is drawn from. */
    publishesNotes: z.boolean(),
  })
  .strict();

/**
 * A module's face: rows of tokens, one per grid cell, as the engine padded them (every row the same
 * length). Equal neighbouring tokens form one rectangular block, as CSS `grid-template-areas`. `.` is
 * an empty cell, `wave` the wave panel, `scope` the scope panel, `value` the readout, `meter` the
 * level meter, `text:<id>` a text property, anything else names a port (`in:`/`out:` when the two
 * sides share an id) or a param (`param:` when it collides with a port). The engine's registry has already
 * checked the geometry; the check here is only that every token still names something on the
 * module, which is what `composeFace` needs to be true.
 */
export const FaceSchema = z.array(z.array(z.string().min(1)).min(1)).min(1);

/** What a face token names on a module, or null for one that names nothing. `.` is `empty`. */
export function resolveFaceToken(
  module: {
    inputs: { id: string }[];
    outputs: { id: string }[];
    params: { id: string }[];
    texts?: { id: string }[];
  },
  token: string,
):
  | { kind: "empty" }
  | { kind: "wave" }
  | { kind: "scope" }
  | { kind: "value" }
  | { kind: "meter" }
  | { kind: "pianoRoll" }
  | { kind: "input" | "output" | "param" | "text"; id: string }
  | null {
  if (token === ".") return { kind: "empty" };
  if (token === "wave") return { kind: "wave" };
  if (token === "scope") return { kind: "scope" };
  if (token === "value") return { kind: "value" };
  if (token === "meter") return { kind: "meter" };
  if (token === "pianoRoll") return { kind: "pianoRoll" };
  // Implicit modulation ports never sit on a face; they ride on their param's control, so only a
  // declared input is a jack.
  const declared = module.inputs.filter(
    (p) => !("implicit" in p && p.implicit === true),
  );
  const has = (list: { id: string }[], id: string) =>
    list.some((p) => p.id === id);
  for (const [prefix, kind, list] of [
    ["in:", "input", declared],
    ["out:", "output", module.outputs],
    ["param:", "param", module.params],
    // A text property is always written out in full. Unlike a port or a param it is not something
    // a bare name could plausibly mean, and spelling it keeps a module with a `pattern` param and
    // a `pattern` string from being ambiguous.
    ["text:", "text", module.texts ?? []],
  ] as const) {
    if (!token.startsWith(prefix)) continue;
    const id = token.slice(prefix.length);
    return has(list, id) ? { kind, id } : null;
  }
  const input = has(declared, token);
  const output = has(module.outputs, token);
  const param = has(module.params, token);
  const matches = Number(input) + Number(output) + Number(param);
  if (matches !== 1) return null;
  if (input) return { kind: "input", id: token };
  if (output) return { kind: "output", id: token };
  return { kind: "param", id: token };
}

export const ModuleDescriptorSchema = z
  .object({
    id: z.string().min(1),
    name: z.string().min(1),
    category: z.string().min(1),
    doc: z.string(),
    flags: ModuleFlagsSchema,
    inputs: z.array(PortDescSchema),
    outputs: z.array(PortDescSchema),
    params: z.array(ParamDescSchema),
    texts: z.array(TextDescSchema),
    /** The declared face, or null for a module that leaves its face to the interface. */
    face: FaceSchema.nullable(),
  })
  .strict()
  .refine(
    (m) =>
      m.face === null ||
      m.face.every((row) => row.every((t) => resolveFaceToken(m, t) !== null)),
    { message: "every face token names a port or a param of the module" },
  )
  .refine(
    (m) =>
      m.face === null ||
      m.face.every((row) => row.length === m.face?.[0].length),
    { message: "a face is a rectangle: every row the same length" },
  );

/** Engine-wide constants a patch is written against; the editor needs them to label pitch and time. */
export const ConventionsSchema = z
  .object({
    octavesPerUnit: z.number(),
    middleCHz: z.number(),
    gateThreshold: z.number(),
    blockSize: z.number().int().positive(),
    lanes: z.array(z.string()).length(4),
  })
  .strict();

export const CatalogSchema = z
  .object({
    /** 64-bit FNV-1a over the module list. The renderer caches by this and refetches when it moves. */
    catalogHash: z.string().regex(/^[0-9a-f]{16}$/),
    conventions: ConventionsSchema,
    modules: z.array(ModuleDescriptorSchema).min(1),
  })
  .strict()
  .refine(
    (c) => c.modules.every((m, i) => i === 0 || c.modules[i - 1].id < m.id),
    { message: "modules must be sorted by id and unique" },
  );

export type PortKind = z.infer<typeof PortKindSchema>;
export type SignalRole = z.infer<typeof SignalRoleSchema>;
export type ParamUnit = z.infer<typeof ParamUnitSchema>;
export type ParamCurve = z.infer<typeof ParamCurveSchema>;
export type PortDesc = z.infer<typeof PortDescSchema>;
export type ParamDesc = z.infer<typeof ParamDescSchema>;
export type TextDesc = z.infer<typeof TextDescSchema>;
export type ModuleDescriptor = z.infer<typeof ModuleDescriptorSchema>;
export type Catalog = z.infer<typeof CatalogSchema>;

/** The input port id the engine generates for a modulatable param. */
export function implicitPortId(paramId: string): string {
  return `param:${paramId}`;
}

export function findModule(
  catalog: Catalog,
  id: string,
): ModuleDescriptor | undefined {
  return catalog.modules.find((m) => m.id === id);
}

File: engine/src/services/Catalog.cpp
#include "services/Catalog.hpp"
#include <algorithm>
#include <cstdint>
#include <string>
#include <vector>
#include "core/Conventions.hpp"

namespace pg {
namespace {

const char* kindName(PortKind k) { return k == PortKind::Event ? "event" : "continuous"; }

const char* roleName(SignalRole r) {
  switch (r) {
    case SignalRole::Any:   return "any";
    case SignalRole::Audio: return "audio";
    case SignalRole::Cv:    return "cv";
    case SignalRole::Gate:  return "gate";
    case SignalRole::Pitch: return "pitch";
    case SignalRole::Phase: return "phase";
    case SignalRole::Note:  return "note";
  }
  return "any";
}

const char* unitName(ParamUnit u) {
  switch (u) {
    case ParamUnit::None:      return "none";
    case ParamUnit::Hz:        return "hz";
    case ParamUnit::Seconds:   return "seconds";
    case ParamUnit::Db:        return "db";
    case ParamUnit::Semitones: return "semitones";
    case ParamUnit::Percent:   return "percent";
    case ParamUnit::Ratio:     return "ratio";
    case ParamUnit::Milliseconds: return "milliseconds";
  }
  return "none";
}

const char* curveName(ParamCurve c) {
  switch (c) {
    case ParamCurve::Linear:  return "linear";
    case ParamCurve::Log:     return "log";
    case ParamCurve::Exp:     return "exp";
    case ParamCurve::Quartic: return "quartic";
  }
  return "linear";
}

/// A descriptor string may be null (the ABI allows it); the JSON must not be.
const char* text(const char* s) { return s ? s : ""; }

nlohmann::json portJson(const PortDesc& p, int32_t paramIndex, const ModuleDescriptor& desc) {
  nlohmann::json j;
  j["id"] = text(p.id);
  j["name"] = text(p.name);
  j["kind"] = kindName(p.kind);
  j["role"] = roleName(p.role);
  j["doc"] = text(p.doc);
  // An implicit port is the one the registry adds for a modulatable param; the UI draws it on the param
  // rather than in the port list, so it has to be able to tell them apart and know which param it feeds.
  j["implicit"] = paramIndex >= 0;
  if (paramIndex >= 0) j["param"] = text(desc.params[paramIndex].id);
  return j;
}

nlohmann::json paramJson(const ParamDesc& p) {
  nlohmann::json j;
  j["id"] = text(p.id);
  j["name"] = text(p.name);
  j["min"] = p.min;
  j["max"] = p.max;
  j["default"] = p.def;
  j["unit"] = unitName(p.unit);
  j["curve"] = curveName(p.curve);
  j["flags"] = {
    {"modulatable", (p.flags & kParamModulatable) != 0},
    {"integer", (p.flags & kParamInteger) != 0},
    {"enum", (p.flags & kParamEnum) != 0},
    {"hidden", (p.flags & kParamHidden) != 0},
    {"noSmooth", (p.flags & kParamNoSmooth) != 0},
    {"structural", (p.flags & kParamStructural) != 0},
    // Whether the module wants this control on its face. The interface shows these without being
    // asked; everything else is reached through the inspector.
    {"primary", (p.flags & kParamPrimary) != 0},
  };
  if (p.enumLabels != nullptr && p.enumCount > 0) {
    std::vector<std::string> labels;
    labels.reserve(p.enumCount);
    for (uint32_t i = 0; i < p.enumCount; ++i) labels.push_back(text(p.enumLabels[i]));
    j["enumLabels"] = labels;
  }
  j["uiWidget"] = text(p.uiWidget);
  if (p.group != nullptr) j["group"] = p.group;
  j["doc"] = text(p.doc);
  return j;
}

/// A text property: the string a module owns and an interface generates an editor for.
nlohmann::json textJson(const TextDesc& t) {
  nlohmann::json j;
  j["id"] = text(t.id);
  j["name"] = text(t.name);
  j["default"] = text(t.def);
  j["flags"] = {{"multiline", (t.flags & kTextMultiline) != 0}};
  // The editor mode to open. Null is plain text, and so is a name the interface has never heard
  // of -- an unknown language should cost syntax colouring, not the ability to type.
  if (t.language != nullptr) j["language"] = text(t.language);
  if (t.placeholder != nullptr) j["placeholder"] = text(t.placeholder);
  j["doc"] = text(t.doc);
  return j;
}

nlohmann::json moduleJson(const RegisteredModule& m) {
  const ModuleDescriptor& d = *m.desc;
  nlohmann::json j;
  j["id"] = text(d.id);
  j["name"] = text(d.name);
  j["category"] = text(d.category);
  j["doc"] = text(d.doc);
  j["flags"] = {
    {"terminal", (d.flags & kModuleTerminal) != 0},
    {"needsTransport", (d.flags & kModuleNeedsTransport) != 0},
    {"writesTelemetry", (d.flags & kModuleWritesTelemetry) != 0},
    {"previewsWave", (d.flags & kModulePreviewsWave) != 0},
    {"previewsEnvelope", (d.flags & kModulePreviewsEnvelope) != 0},
    {"publishesScope", (d.flags & kModulePublishesScope) != 0},
    {"publishesValue", (d.flags & kModulePublishesValue) != 0},
    {"publishesMeter", (d.flags & kModulePublishesMeter) != 0},
    {"publishesNotes", (d.flags & kModulePublishesNotes) != 0},
    {"publishesKeys", (d.flags & kModulePublishesKeys) != 0},
    {"voiceEntry", (d.flags & kModuleVoiceEntry) != 0},
    {"voiceExit", (d.flags & kModuleVoiceExit) != 0},
  };
  // The registry's input list is the declared ports followed by one implicit port per modulatable param,
  // in exactly the order the compiler assigns buffers, so the UI's port indices match the engine's.
  nlohmann::json inputs = nlohmann::json::array();
  for (size_t i = 0; i < m.inputs.size(); ++i) inputs.push_back(portJson(m.inputs[i], m.inputParam[i], d));
  j["inputs"] = inputs;
  nlohmann::json outputs = nlohmann::json::array();
  for (uint32_t i = 0; i < d.numOutputs; ++i) outputs.push_back(portJson(d.outputs[i], -1, d));
  j["outputs"] = outputs;
  nlohmann::json params = nlohmann::json::array();
  for (uint32_t i = 0; i < d.numParams; ++i) params.push_back(paramJson(d.params[i]));
  j["params"] = params;
  nlohmann::json texts = nlohmann::json::array();
  for (uint32_t i = 0; i < d.numTexts; ++i) texts.push_back(textJson(d.texts[i]));
  j["texts"] = texts;
  // The face as the registry padded it: rows of tokens, every row the same length, or null for a module
  // that leaves its face to the interface. See `ModuleDescriptor::face`.
  if (m.face.empty()) j["face"] = nullptr;
  else j["face"] = m.face;
  return j;
}

/// 64-bit FNV-1a over the serialized modules. Content-addressed on purpose: the renderer caches the catalog
/// and only refetches when the hash moves, so it has to change whenever anything a module draws with does.
std::string fnv1a(const std::string& bytes) {
  uint64_t hash = 1469598103934665603ULL;
  for (const char c : bytes) {
    hash ^= static_cast<uint8_t>(c);
    hash *= 1099511628211ULL;
  }
  static const char* kHex = "0123456789abcdef";
  std::string out(16, '0');
  for (int i = 15; i >= 0; --i) {
    out[static_cast<size_t>(i)] = kHex[hash & 0xf];
    hash >>= 4;
  }
  return out;
}

}  // namespace

nlohmann::json catalogJson(const Registry& registry) {
  std::vector<const RegisteredModule*> modules = registry.all();
  std::sort(modules.begin(), modules.end(),
            [](const RegisteredModule* a, const RegisteredModule* b) { return std::string(a->desc->id) < b->desc->id; });

  nlohmann::json list = nlohmann::json::array();
  for (const RegisteredModule* m : modules) list.push_back(moduleJson(*m));

  nlohmann::json out;
  out["catalogHash"] = fnv1a(list.dump());
  // Written as doubles rather than promoted from the engine's floats: `static_cast<double>(261.6256f)` is
  // 261.62560272216797, which is the same pitch but an unreadable document. The asserts keep the two in step.
  static_assert(kMiddleCHz == 261.6256f);
  static_assert(kOctavesPerUnit == 10.f);
  out["conventions"] = {
    {"octavesPerUnit", 10.0},
    {"middleCHz", 261.6256},
    {"gateThreshold", 0},
    {"blockSize", kMaxBlockSize},
    {"lanes", {"v0.L", "v0.R", "v1.L", "v1.R"}},
  };
  out["modules"] = std::move(list);
  return out;
}

}  // namespace pg
